python/pyarrow/_flight.pyx

```diff
@@ -1667,6 +1812,258 @@ cdef class ClientAuthHandler:
+
+    All middleware methods will be called from the same thread as the
+    RPC method implementation. That is, thread-locals set in the
+    middleware are accessible from the method itself.
+
+    """
+
+    def start_call(self, info, headers):
+        """Called at the start of an RPC.
+
+        This must be thread-safe.
+
+        Parameters
+        ----------
+        info : CallInfo
+            Information about the call.
+
+        headers : dict
+            A dictionary of headers from the client. Keys are strings
+            and values are lists of strings.
+
+        Returns
+        -------
+        instance : ServerMiddleware
+            An instance of ServerMiddleware (the instance to use for
+            the call), or None if this call is not intercepted.
+
+        Raises
+        ------
+        exception : pyarrow.ArrowException
+            If an exception is raised, the call will be rejected with
+            the given error.
+
+        """
+
+
+cdef class ServerMiddleware:
+    """Server-side middleware for a call, instantiated per RPC.
+
+    Methods here should be fast and must be infalliable: they should
+    not raise exceptions or stall indefinitely.
+
+    """
+
+    def sending_headers(self):
+        """A callback before headers are sent.
+
+        Returns
+        -------
+        headers : dict
+            A dictionary of header values to add to the response, or
+            None if no headers are to be added. The dictionary should
+            have string keys and string or list-of-string values.
+
+        """
+
+    def call_completed(self, exception):
+        """A callback when the call finishes.
+
+        Parameters
+        ----------
+        exception : pyarrow.ArrowException
+            If the call errored, this is the equivalent
+            exception. Will be None if the call succeeded.
+
+        """
+
+    @staticmethod
+    cdef void wrap(object py_middleware,
+                   shared_ptr[CServerMiddleware]* c_instance):
+        cdef PyServerMiddlewareVtable vtable
+        vtable.sending_headers = _middleware_sending_headers
+        vtable.call_completed = _middleware_call_completed
+        c_instance[0].reset(new CPyServerMiddleware(py_middleware, vtable))
+
+
+cdef class _ServerMiddlewareFactoryWrapper(ServerMiddlewareFactory):
+    """A wrapper that bundles server middleware into a single C++ one.
+
+    gRPC runs RPCs on C++ threads, so we use PyGilState_Ensure/Release
+    to interact with the GIL before running Python code. However, this
+    has a major shortcoming: it breaks thread locals. When we enter
+    Python, it creates new thread state. When we exit, that thread
+    state is freed (and Python thread locals cleaned up), since the
+    thread isn't persistent from Python's point of view. This makes
+    development of middleware harder than it needs to be.
+
+    This middleware solves this by first incrementing the reference
```

Comment:
I don't understand the concern. Using thread locals is a low-level way of looking at it. Instead people should probably keep state on their middleware instances.

Reply:
What this tries to solve is sharing state between the middleware and the rest of the application. (A specific application would be getting a request ID out of the incoming headers and automatically sending them in outgoing headers.) For Java, I went with providing a getter for middleware instances based on a user-supplied key, which is messy, but is required since gRPC makes no guarantees about threading in Java. I could do the same here, and add a fake type ID mechanism like we did for extension types, if that's a more reliable solution.

Reply:
Can you give an example of "sharing state"? I don't understand why that's not possible by e.g. passing state through the middleware factory.

Reply:
Also, the threading guarantees would be gRPC-specific, other backends may not offer the same guarantees.

Reply:
> Also, the threading guarantees would be gRPC-specific, other backends may not offer the same guarantees.

That's true. I will fall back on the approach I took for Java, then.

> Can you give an example of "sharing state"? I don't understand why that's not possible by e.g. passing state through the middleware factory.

Passing it through the middleware factory would mean that you need to somehow identify what state corresponds to the current thread of execution.

The specific goal is to enable something like [OpenTracing](https://opentracing.io/docs/overview/). When a request comes in, it may have a request ID and other metadata attached in headers. An OpenTracing middleware would take this and (somehow) stash it locally such that any subsequent requests made by the server while handling this request, whether REST, gRPC, or Flight, would propagate the same request ID. Ideally, this would also be transparent to the application code: it would make requests as normal, and the library-specific client middleware would take care of finding that context and attaching it to outgoing requests.

Often, this is done with thread-locals, or with specific integrations with `asyncio` or other libraries: https://opentracing.io/guides/python/#scope-managers

Reply:
Hmm. Note that you can emulate thread locals simply with a global dict indexed by thread id. You need to take care about the lifetime of the dict values to avoid a memory leak, but the middleware tells you when a request is finished.

I don't think I understand how the Java solution works, but I guess I'll see it in action once you update this PR :-)